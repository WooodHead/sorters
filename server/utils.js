function prepare(o) {
    if (o && o._id) {
        o._id = o._id.toString()
    }
    return o
}

async function findOne(Collection, ...args) {
    return prepare(await Collection.findOne(...args))
}

async function find(Collection, ...args) {
    return (await Collection.find(...args).toArray()).map(prepare)
}

module.exports = {
    prepare,
    findOne,
    find,
}