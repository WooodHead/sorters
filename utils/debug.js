
export function log(arg) {
    console.log(arg)
    return arg
}
