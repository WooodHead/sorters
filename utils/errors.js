export const errorMessage = (e) => {
    if (e.graphQLErrors && e.graphQLErrors.length > 0) {
        return e.graphQLErrors[0].message
    } else {
        return e.message
    }
}