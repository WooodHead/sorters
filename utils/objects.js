
function toObject(arr) {
    const res = {}
    for (const [key, value] of arr) {
        res[key] = value
    }
    return res
}

function objectMap(object, keyMap = key => key, valueMap = value => value) {
    return Object.keys(object).reduce((o, key) => {
        o[keyMap(key)] = valueMap(object[key])
        return o
    }, {})
}

module.exports = {
    toObject,
    objectMap
}
