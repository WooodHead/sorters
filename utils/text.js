export const replaceAll = (string, search, replacement) => {
  if (typeof string === "string") {
    return string.replace(new RegExp(search, 'g'), replacement)
  } else {
    return string
  }
}

export const teaserAndInfo = (content, min, opt, max) => {
	if (content) {
		if (content.length <= max) {
			return {
				cut: false,
				teaser: content
			};
		}
		const words = content.split(" ");
		let teaser = "";
		for (let i=0; i<words.length; i++) {
			const candidate = words[i];
			let wouldbelength = teaser.length + 1 + candidate.length + 3;
			if (wouldbelength >= opt) {
				// We're done
				if (wouldbelength > max) {
					if (teaser.length + 3 < min) {
						// cut the word
						return {
							teaser: teaser + " " + candidate.substring(0, opt - teaser.length - 1) + "...",
							cut: true,
						};
					} else {
						return {
							teaser: teaser + "...",
							cut: true,
						};
					}
				} else {
					return {
						teaser: teaser + " " + candidate + "...",
						cut: true,
					};
				}
			}
			teaser += " " + candidate;
		}
		return {
			teaser: teaser + "...",
			cut: true,
		};
	} else {
		return {
			teaser: "",
			cut: false,
		};
	}
};
export const teaser = (content, min, opt, max) => {
	const teaser = teaserAndInfo(content, min, opt, max);
	return teaser.teaser;
}
