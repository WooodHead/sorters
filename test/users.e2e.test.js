import cheerio from 'cheerio'
import {pretty} from './utils'
import {MongoClient, ObjectID} from 'mongodb'
import {setup, teardown} from './setup'
import {generateAndLogUser, setUserData} from './fixtures'

let db
let browser

beforeAll(async () => {
    const res = await setup()
    db = res.db
    browser = res.browser
})

afterEach(async () => {
    await db.dropDatabase()    
})

afterAll(async () => {
    await teardown({db, browser})
})

describe('users', () => {
    it('displays', async () => {
        const browserPage = await browser.createPage()
        const status = await browserPage.open(`http://localhost:3000/users`)
        expect(status).toBe('success')

        const text = await browserPage.property('content')
        const $ = cheerio.load(text)
        const page = $('#__next')
        const html = pretty(page.html())
        expect(html).toMatchSnapshot()
    })

    it('displays a user', async () => {        
        const Users = db.collection('users')
        const user = await Users.insertOne({
            local: {
                username: 'test'
            }
        })

        const browserPage = await browser.createPage()
        const status = await browserPage.open(`http://localhost:3000/users`)
        expect(status).toBe('success')

        const text = await browserPage.property('content')
        const $ = cheerio.load(text)
        const page = $('#__next')
        const html = pretty(page.html())
        expect(html).toMatchSnapshot()
    })

    it('displays a user with data', async () => {        
        const browserPage = await browser.createPage()

        await generateAndLogUser(browserPage)
        await setUserData(browserPage)

        const status = await browserPage.open(`http://localhost:3000/users`)
        expect(status).toBe('success')

        const text = await browserPage.property('content')
        const $ = cheerio.load(text)
        const page = $('#__next')
        const html = pretty(page.html())
        expect(html).toMatchSnapshot()
    })

})
