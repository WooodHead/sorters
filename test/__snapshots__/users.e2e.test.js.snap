// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users displays 1`] = `
"<div data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"1089078876\\">
  <div data-reactid=\\"2\\">
    <!-- react-empty: 3 -->
    <div style=\\"padding-top:50px;padding-bottom:20px;\\" data-reactid=\\"4\\">
      <nav class=\\"navbar navbar-inverse navbar-fixed-top\\" data-reactid=\\"5\\">
        <div class=\\"container\\" data-reactid=\\"6\\">
          <div class=\\"navbar-header\\" data-reactid=\\"7\\">
            <button type=\\"button\\" class=\\"navbar-toggle collapsed\\" data-toggle=\\"collapse\\" data-target=\\"#navbar\\" aria-expanded=\\"false\\" aria-controls=\\"navbar\\" data-reactid=\\"8\\"><span class=\\"sr-only\\" data-reactid=\\"9\\">Toggle navigation</span><span class=\\"icon-bar\\" data-reactid=\\"10\\"></span><span class=\\"icon-bar\\" data-reactid=\\"11\\"></span><span class=\\"icon-bar\\" data-reactid=\\"12\\"></span></button>
            <a class=\\"navbar-brand\\"
              href=\\"/\\" data-reactid=\\"13\\">Sorters Club</a>
          </div>
          <div id=\\"navbar\\" class=\\"collapse navbar-collapse\\" data-reactid=\\"14\\">
            <ul class=\\"nav navbar-nav navbar-right\\" data-reactid=\\"15\\">
              <li class=\\"false\\" data-reactid=\\"16\\">
                <a href=\\"/\\" data-reactid=\\"17\\">Home</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"18\\">
                <a href=\\"/about\\" data-reactid=\\"19\\">About</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"20\\">
                <a href=\\"/news\\" data-reactid=\\"21\\">News</a>
              </li>
              <li class=\\"active\\" data-reactid=\\"22\\">
                <a href=\\"/users\\" data-reactid=\\"23\\">Sorters</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"24\\">
                <a href=\\"/dashboard\\" data-reactid=\\"25\\">Dashboard</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"26\\">
                <a href=\\"/login\\" data-reactid=\\"27\\">Log in</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"28\\">
                <a href=\\"/register\\" data-reactid=\\"29\\">Register</a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </div>
    <div class=\\"container\\" data-reactid=\\"30\\">
      <h1 data-reactid=\\"31\\">Sorters</h1>
      <p data-reactid=\\"32\\">Here is a list of all the sorters with a public profile.</p>
      <div class=\\"row\\" data-reactid=\\"33\\"></div>
    </div>
    <footer data-reactid=\\"34\\">
      <div class=\\"container\\" data-reactid=\\"35\\">
        <hr data-reactid=\\"36\\">
        <p data-reactid=\\"37\\">
          <!-- react-text: 38 -->Brought to you with &#x2764; by
          <!-- /react-text -->
          <a href=\\"/about\\" data-reactid=\\"39\\">Nick Redmark</a>
          <!-- react-text: 40 -->. Support and discuss the development of this platform on
          <!-- /react-text -->
          <a href=\\"https://www.patreon.com/nickredmark\\" target=\\"_blank\\" data-reactid=\\"41\\">Patreon</a>
          <!-- react-text: 42 -->.
          <!-- /react-text -->
        </p>
      </div>
    </footer>
  </div>
</div>"
`;

exports[`users displays a user 1`] = `
"<div data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"199424523\\">
  <div data-reactid=\\"2\\">
    <!-- react-empty: 3 -->
    <div style=\\"padding-top:50px;padding-bottom:20px;\\" data-reactid=\\"4\\">
      <nav class=\\"navbar navbar-inverse navbar-fixed-top\\" data-reactid=\\"5\\">
        <div class=\\"container\\" data-reactid=\\"6\\">
          <div class=\\"navbar-header\\" data-reactid=\\"7\\">
            <button type=\\"button\\" class=\\"navbar-toggle collapsed\\" data-toggle=\\"collapse\\" data-target=\\"#navbar\\" aria-expanded=\\"false\\" aria-controls=\\"navbar\\" data-reactid=\\"8\\"><span class=\\"sr-only\\" data-reactid=\\"9\\">Toggle navigation</span><span class=\\"icon-bar\\" data-reactid=\\"10\\"></span><span class=\\"icon-bar\\" data-reactid=\\"11\\"></span><span class=\\"icon-bar\\" data-reactid=\\"12\\"></span></button>
            <a class=\\"navbar-brand\\"
              href=\\"/\\" data-reactid=\\"13\\">Sorters Club</a>
          </div>
          <div id=\\"navbar\\" class=\\"collapse navbar-collapse\\" data-reactid=\\"14\\">
            <ul class=\\"nav navbar-nav navbar-right\\" data-reactid=\\"15\\">
              <li class=\\"false\\" data-reactid=\\"16\\">
                <a href=\\"/\\" data-reactid=\\"17\\">Home</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"18\\">
                <a href=\\"/about\\" data-reactid=\\"19\\">About</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"20\\">
                <a href=\\"/news\\" data-reactid=\\"21\\">News</a>
              </li>
              <li class=\\"active\\" data-reactid=\\"22\\">
                <a href=\\"/users\\" data-reactid=\\"23\\">Sorters</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"24\\">
                <a href=\\"/dashboard\\" data-reactid=\\"25\\">Dashboard</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"26\\">
                <a href=\\"/login\\" data-reactid=\\"27\\">Log in</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"28\\">
                <a href=\\"/register\\" data-reactid=\\"29\\">Register</a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </div>
    <div class=\\"container\\" data-reactid=\\"30\\">
      <h1 data-reactid=\\"31\\">Sorters</h1>
      <p data-reactid=\\"32\\">Here is a list of all the sorters with a public profile.</p>
      <div class=\\"row\\" data-reactid=\\"33\\">
        <div class=\\"col-xs-12 col-md-6\\" style=\\"display:flex;margin-bottom:24px;clear:both;\\" data-reactid=\\"34\\">
          <div data-reactid=\\"35\\">
            <a href=\\"/u/test\\" data-reactid=\\"36\\">
              <img alt=\\"Gravatar for undefined\\" style=\\"margin-right:24px;\\" src=\\"//www.gravatar.com/avatar/test?d=retro&amp;r=g&amp;s=100\\" srcset=\\"//www.gravatar.com/avatar/test?d=retro&amp;r=g&amp;s=200 2x\\" height=\\"100\\" width=\\"100\\" class=\\"react-gravatar\\"
                data-reactid=\\"37\\">
            </a>
          </div>
          <div data-reactid=\\"38\\">
            <h3 style=\\"margin-top:0;\\" data-reactid=\\"39\\">
              <a href=\\"/u/test\\" data-reactid=\\"40\\">test</a>
            </h3>
            <a href=\\"/u/test\\" data-reactid=\\"41\\">
              <!-- react-text: 42 -->/u/
              <!-- /react-text -->
              <!-- react-text: 43 -->test
              <!-- /react-text -->
            </a>
            <div data-reactid=\\"44\\">
              <div data-reactid=\\"45\\"></div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <footer data-reactid=\\"46\\">
      <div class=\\"container\\" data-reactid=\\"47\\">
        <hr data-reactid=\\"48\\">
        <p data-reactid=\\"49\\">
          <!-- react-text: 50 -->Brought to you with &#x2764; by
          <!-- /react-text -->
          <a href=\\"/about\\" data-reactid=\\"51\\">Nick Redmark</a>
          <!-- react-text: 52 -->. Support and discuss the development of this platform on
          <!-- /react-text -->
          <a href=\\"https://www.patreon.com/nickredmark\\" target=\\"_blank\\" data-reactid=\\"53\\">Patreon</a>
          <!-- react-text: 54 -->.
          <!-- /react-text -->
        </p>
      </div>
    </footer>
  </div>
</div>"
`;

exports[`users displays a user with data 1`] = `
"<div data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"346684231\\">
  <div data-reactid=\\"2\\">
    <!-- react-empty: 3 -->
    <div style=\\"padding-top:50px;padding-bottom:20px;\\" data-reactid=\\"4\\">
      <nav class=\\"navbar navbar-inverse navbar-fixed-top\\" data-reactid=\\"5\\">
        <div class=\\"container\\" data-reactid=\\"6\\">
          <div class=\\"navbar-header\\" data-reactid=\\"7\\">
            <button type=\\"button\\" class=\\"navbar-toggle collapsed\\" data-toggle=\\"collapse\\" data-target=\\"#navbar\\" aria-expanded=\\"false\\" aria-controls=\\"navbar\\" data-reactid=\\"8\\"><span class=\\"sr-only\\" data-reactid=\\"9\\">Toggle navigation</span><span class=\\"icon-bar\\" data-reactid=\\"10\\"></span><span class=\\"icon-bar\\" data-reactid=\\"11\\"></span><span class=\\"icon-bar\\" data-reactid=\\"12\\"></span></button>
            <a class=\\"navbar-brand\\"
              href=\\"/\\" data-reactid=\\"13\\">Sorters Club</a>
          </div>
          <div id=\\"navbar\\" class=\\"collapse navbar-collapse\\" data-reactid=\\"14\\">
            <ul class=\\"nav navbar-nav navbar-right\\" data-reactid=\\"15\\">
              <li class=\\"false\\" data-reactid=\\"16\\">
                <a href=\\"/\\" data-reactid=\\"17\\">Home</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"18\\">
                <a href=\\"/about\\" data-reactid=\\"19\\">About</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"20\\">
                <a href=\\"/news\\" data-reactid=\\"21\\">News</a>
              </li>
              <li class=\\"active\\" data-reactid=\\"22\\">
                <a href=\\"/users\\" data-reactid=\\"23\\">Sorters</a>
              </li>
              <li class=\\"false\\" data-reactid=\\"24\\">
                <a href=\\"/dashboard\\" data-reactid=\\"25\\">Dashboard</a>
              </li>
              <li class=\\"dropdownfalse\\" data-reactid=\\"26\\">
                <a href=\\"#\\" class=\\"dropdown-toggle\\" data-toggle=\\"dropdown\\" role=\\"button\\" aria-haspopup=\\"true\\" aria-expanded=\\"true\\" data-reactid=\\"27\\">
                  <!-- react-text: 28 -->Hello,
                  <!-- /react-text -->
                  <!-- react-text: 29 -->test
                  <!-- /react-text -->
                  <!-- react-text: 30 -->
                  <!-- /react-text --><span class=\\"caret\\" data-reactid=\\"31\\"></span></a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </div>
    <div class=\\"container\\" data-reactid=\\"32\\">
      <h1 data-reactid=\\"33\\">Sorters</h1>
      <p data-reactid=\\"34\\">Here is a list of all the sorters with a public profile.</p>
      <div class=\\"row\\" data-reactid=\\"35\\">
        <div class=\\"col-xs-12 col-md-6\\" style=\\"display:flex;margin-bottom:24px;clear:both;\\" data-reactid=\\"36\\">
          <div data-reactid=\\"37\\">
            <a href=\\"/u/test\\" data-reactid=\\"38\\">
              <img alt=\\"Gravatar for undefined\\" style=\\"margin-right:24px;\\" src=\\"//www.gravatar.com/avatar/23463b99b62a72f26ed677cc556c44e8?d=retro&amp;r=g&amp;s=100\\" srcset=\\"//www.gravatar.com/avatar/23463b99b62a72f26ed677cc556c44e8?d=retro&amp;r=g&amp;s=200 2x\\"
                height=\\"100\\" width=\\"100\\" class=\\"react-gravatar\\" data-reactid=\\"39\\">
            </a>
          </div>
          <div data-reactid=\\"40\\">
            <h3 style=\\"margin-top:0;\\" data-reactid=\\"41\\">
              <a href=\\"/u/test\\" data-reactid=\\"42\\">Test User</a>
            </h3>
            <a href=\\"/u/test\\" data-reactid=\\"43\\">
              <!-- react-text: 44 -->/u/
              <!-- /react-text -->
              <!-- react-text: 45 -->test
              <!-- /react-text -->
            </a>
            <div data-reactid=\\"46\\">
              <div class=\\"markdown\\" data-reactid=\\"47\\">
                <p>Test user about</p>
              </div>
            </div>
            <div data-reactid=\\"48\\">
              <!-- react-text: 49 -->&#x1F4D6;
              <!-- /react-text -->
              <!-- react-text: 50 -->3
              <!-- /react-text -->
              <!-- react-text: 51 -->
              <!-- /react-text -->
              <a href=\\"/u/test/reads\\" data-reactid=\\"52\\">books</a><span data-reactid=\\"53\\"><!-- react-text: 54 -->, reading <!-- /react-text --><em data-reactid=\\"55\\">Read - reading</em></span></div>
            <div data-reactid=\\"56\\">
              <!-- react-text: 57 -->&#x25CE;
              <!-- /react-text -->
              <!-- react-text: 58 -->3
              <!-- /react-text -->
              <!-- react-text: 59 -->
              <!-- /react-text -->
              <a href=\\"/u/test/goals\\" data-reactid=\\"60\\">goals</a><span data-reactid=\\"61\\"><!-- react-text: 62 -->, working on <!-- /react-text --><em data-reactid=\\"63\\">Goal - doing</em></span></div>
          </div>
        </div>
      </div>
    </div>
    <footer data-reactid=\\"64\\">
      <div class=\\"container\\" data-reactid=\\"65\\">
        <hr data-reactid=\\"66\\">
        <p data-reactid=\\"67\\">
          <!-- react-text: 68 -->Brought to you with &#x2764; by
          <!-- /react-text -->
          <a href=\\"/about\\" data-reactid=\\"69\\">Nick Redmark</a>
          <!-- react-text: 70 -->. Support and discuss the development of this platform on
          <!-- /react-text -->
          <a href=\\"https://www.patreon.com/nickredmark\\" target=\\"_blank\\" data-reactid=\\"71\\">Patreon</a>
          <!-- react-text: 72 -->.
          <!-- /react-text -->
        </p>
      </div>
    </footer>
  </div>
</div>"
`;
